### PR TITLE
ipn: fix incorrect handling of packet filter changes

### PR DESCRIPTION
**Pull request by danderson**
Fixes a race condition where the packet filter may reject all traffic
to a device's Tailscale IP.

Signed-off-by: David Anderson <danderson@tailscale.com>

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/630)

### CODE REVIEW

**Comment by bradfitz** [comment]
Can you describe the series of events where the problem occurred?

File: ipn/local.go [hunk 1]
```diff
@@ -439,9 +439,15 @@ func (b *LocalBackend) Start(opts Options) error {
 // updateFilter updates the packet filter in wgengine based on the
 // given netMap and user preferences.
 func (b *LocalBackend) updateFilter(netMap *controlclient.NetworkMap, prefs *Prefs) {
+	// NOTE(danderson): if you return early before the
+	// deepprint.UpdateHash() change detection logic, and also change
+	// the fitler (e.g. to block everything during a reconfig), make
```
Comment (bradfitz) [inline ipn/local.go:444]: typo: filter

File: ipn/local.go [hunk 2]
```diff
@@ -454,7 +460,7 @@ func (b *LocalBackend) updateFilter(netMap *controlclient.NetworkMap, prefs *Pre
 	// Be conservative while not ready.
 	shieldsUp := prefs == nil || prefs.ShieldsUp
 
-	changed := deepprint.UpdateHash(&b.filterHash, packetFilter, advRoutes, shieldsUp)
+	changed := deepprint.UpdateHash(&b.filterHash, netMap.Addresses, packetFilter, advRoutes, shieldsUp)
```
Comment (bradfitz) [inline ipn/local.go:463]: Or move this after the UpdateHash so there's no early return. You'll need to do something like:

```go
haveNetmap := netMap != nil
var addesses ...
var packetFilter ..
if haveNetmap {
   addresses = netMap.Addresses
   packetFilter = ...
}
changed := UpdateHash(&b.filterHash, haveNetMap, addresses, packetFilter, advRoutes, shieldsUp)
```

Comment (bradfitz) [reply ipn/local.go:463]: (idea being then you don't need a big scary comment at top and there's one fewer way out of the func)

Comment (danderson) [reply ipn/local.go]: Like this?

**Pull request merged**